### PR TITLE
Projection: allow projections on trace spaces

### PR DESCRIPTION
**Pull request by sv2518**
I want to be able to project functions into the HDiv trace space. This works by solving something like `\int_{\Gamma} \gamma \lambda \diff S = \int_{\Gamma} \gamma \lambda_0 \diff S, \forall \gamma \in T` where `lambda_0` is what we project (kudos to @colinjcotter).

The variational problem changes if we are on an extruded mesh or not. I am only testing this for tets so that I can use interpolating to get a reference solution, but I could set up a reference solver if we want to tests for quads too. In the calculation of the norm I am multiplying by the area,  so that it doesn't diverge as the mesh size goes to zero (also kudos to @colinjcotter).

If you can think of cases where the math behind this doesn't work I can include assertions.

### CODE REVIEW

File: firedrake/projection.py [hunk 1]
```diff
@@ -163,7 +177,21 @@ class BasicProjector(ProjectorBase):
...
         v = firedrake.TestFunction(self.target.function_space())
-        form = firedrake.inner(self.source, v)*firedrake.dx
+        F = self.target.function_space()
+        mixed = isinstance(F.ufl_element(), ufl.MixedElement)
+        if not mixed and isinstance(F.finat_element, HDivTrace):
+            if F.extruded:
+                form = (firedrake.inner(self.source, v)*firedrake.ds_t
+                        + firedrake.inner(self.source, v)*firedrake.ds_v
+                        + firedrake.inner(self.source, v)*firedrake.ds_b
+                        + firedrake.inner(self.source('+'), v('+'))*firedrake.dS_h
```
Comment (dham) [inline firedrake/projection.py:187]: Average of source because it could legitimately be discontinuous.

File: firedrake/projection.py [hunk 2]
```diff
@@ -163,7 +177,21 @@ class BasicProjector(ProjectorBase):
...
+                        + firedrake.inner(self.source('+'), v('+'))*firedrake.dS_v)
+            else:
+                form = (firedrake.inner(self.source, v)*firedrake.ds
+                        + firedrake.inner(self.source('+'), v('+'))*firedrake.dS)
```
Comment (dham) [inline firedrake/projection.py:191]: also average source here.

**Review by dham — CHANGES_REQUESTED** [review]
Please document what this does in the `project` and/or `Projector` documentation.

File: firedrake/projection.py (hunk 2 above)
Comment (sv2518) [reply firedrake/projection.py:191]: We probably want to change this here 
https://github.com/firedrakeproject/firedrake/blob/1fafb01f15f817bece2cbabddd1367b2db7bc151/tests/extrusion/test_trace_extr.py#L24 and here https://github.com/firedrakeproject/firedrake/blob/1fafb01f15f817bece2cbabddd1367b2db7bc151/tests/regression/test_trace_galerkin_projection.py#L59 too?

Comment (sv2518) [reply firedrake/projection.py:191]: Ah, actually I can dismiss this, both tests define the source in a continuous space.

**Review by dham — APPROVED** [review]

**Pull request merged**